platforms/allwinner-d1/boards/lichee-rv/src/main.rs

```diff
@@ -0,0 +1,263 @@
+
+        plic.activate(Interrupt::DMAC_NS, Priority::P1).unwrap();
+        plic.activate(Interrupt::UART0, Priority::P1).unwrap();
+    }
+
+    timer0.start_counter(0xFFFF_FFFF);
+
+    loop {
+        // Tick the scheduler
+        let start = timer0.current_value();
+        let tick = k.tick();
+
+        // Timer is downcounting
+        let elapsed = start.wrapping_sub(timer0.current_value());
+        let turn = k.timer().force_advance_ticks(elapsed.into());
+
+        // If there is nothing else scheduled, and we didn't just wake something up,
+        // sleep for some amount of time
+        if !tick.has_remaining && turn.expired != 0 {
+            let wfi_start = timer0.current_value();
+
+            // TODO(AJM): Sometimes there is no "next" in the timer wheel, even though there should
+            // be. Don't take lack of timer wheel presence as the ONLY heuristic of whether we
+            // should just wait for SOME interrupt to occur. For now, force a max sleep of 100ms
+            // which is still probably wrong.
+            let amount = turn.ticks_to_next_deadline().unwrap_or(100 * 1000 * 3); // 3 ticks per us, 1000 us per ms, 100ms sleep
+
+            // Don't sleep for too long until james figures out wrapping timers
+            let amount = amount.min(0x4000_0000) as u32;
+            let _ = timer1.get_and_clear_interrupt();
+            unsafe {
+                plic.activate(Interrupt::TIMER1, Priority::P1).unwrap();
+            }
+            timer1.set_interrupt_en(true);
+            timer1.start_counter(amount);
+
+            unsafe {
+                riscv::asm::wfi();
+            }
+            // Disable the timer interrupt in case that wasn't what woke us up
+            plic.deactivate(Interrupt::TIMER1).unwrap();
+            timer1.set_interrupt_en(false);
+            timer1.stop();
+
+            // Account for time slept
+            let elapsed = wfi_start.wrapping_sub(timer0.current_value());
+            let _turn = k.timer().force_advance_ticks(elapsed.into());
+        }
+    }
+}
+
+// We don't actually do anything in the TIMER1 interrupt. It is only here to
+// knock us out of WFI. Just disable the IRQ to prevent refires
+fn timer1_int() {
+    let timer = unsafe { &*TIMER::PTR };
+    timer
+        .tmr_irq_sta
+        .modify(|_r, w| w.tmr1_irq_pend().set_bit());
+
+    // Wait for the interrupt to clear to avoid repeat interrupts
+    while timer.tmr_irq_sta.read().tmr1_irq_pend().bit_is_set() {}
+}
+
+fn initialize_kernel() -> Result<&'static Kernel, ()> {
+    let k_settings = KernelSettings {
+        heap_start: AHEAP.as_ptr(),
+        heap_size: HEAP_SIZE,
+        max_drivers: 16,
+        k2u_size: 4096,
+        u2k_size: 4096,
+        timer_granularity: Duration::from_nanos(333),
+    };
+    let k = unsafe { Kernel::new(k_settings).map_err(drop)?.leak().as_ref() };
+    Ok(k)
+}
+
+#[panic_handler]
+fn handler(info: &PanicInfo) -> ! {
```

Comment:
would eventually be nice if the kernel could provide a cross platform panic hook with nice formatting, but that should be a later thing...